sessionctx/variable/noop.go

```diff
@@ -23,6 +23,39 @@ import (
 // but changing them has no effect on behavior.
 
 var noopSysVars = []*SysVar{
+	// It is unsafe to pretend that any variation of "read only" is enabled when the server
+	// does not support it. It is possible that these features will be supported in future,
+	// but until then...
+	{Scope: ScopeGlobal | ScopeSession, Name: TxReadOnly, Value: BoolOff, Type: TypeBool, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		if TiDBOptOn(normalizedValue) && !vars.EnableNoopFuncs {
+			return BoolOff, ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
+		}
+		return normalizedValue, nil
+	}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TransactionReadOnly, Value: BoolOff, Type: TypeBool, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		if TiDBOptOn(normalizedValue) && !vars.EnableNoopFuncs {
+			return BoolOff, ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
+		}
+		return normalizedValue, nil
+	}},
+	{Scope: ScopeGlobal, Name: OfflineMode, Value: BoolOff, Type: TypeBool, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		if TiDBOptOn(normalizedValue) && !vars.EnableNoopFuncs {
+			return BoolOff, ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
```

Comment:
Maybe it's not related to READ ONLY?

Reply:
Offline mode is like read-only, but it also terminates connections from non-SUPER/CONNECTION_ADMIN users.  It was introduced in MySQL 5.7, since "read only" is unsafe in protecting against stale reads. I can change the error to OFFLINE_MODE specifically though.